File: java/src/main/java/com/codurance/training/tasks/service/TaskService.java
package com.codurance.training.tasks.service;

import com.codurance.training.tasks.Task;

import java.util.ArrayList;
import java.util.Collection;
import java.util.List;
import java.util.Map;

public class TaskService {

    private TaskRepository taskRepository;

    public TaskService(TaskRepository taskRepository) {
        this.taskRepository = taskRepository;
    }

    public void addProject(String projectName) {
        taskRepository.getTasks().put(projectName, new ArrayList<Task>());
    }

    public Collection<String> findAllProjects() {
        return taskRepository.getTasks().keySet();
    }

    public Collection<Task> findAllTasksForProject(String projectName) {
        Map<String, List<Task>> projectNameToTasksMap = taskRepository.getTasks();
        return projectNameToTasksMap.get(projectName);
    }

    public ActionResult addTaskToProject(String project, String description) {
        if (taskRepository.projectWithNameExists(project)) {
            Collection<Task> projectTasks = findAllTasksForProject(project);
            Task task = new Task(taskRepository.nextId(), description, false);
            projectTasks.add(task);
            return new ActionSuccessful(task.getId());
        }
        return new ActionFailed();
    }

    public ActionResult setTaskDone(long id, boolean done) {
        for (Map.Entry<String, List<Task>> project : taskRepository.getTasks().entrySet()) {
            for (Task task : project.getValue()) {
                if (task.getId() == id) {
                    task.setDone(done);
                    return new ActionSuccessful(id);
                }
            }
        }
        return new ActionFailed();
    }

    // TODO add missing test for: no projects, no tasks in project, only other task in project
    public Task findTaskById(long taskId) {
        for (Map.Entry<String, List<Task>> project : taskRepository.getTasks().entrySet()) {
            for (Task task : project.getValue()) {
                if (task.getId() == taskId) {
                    return task;
                }
            }
        }
        return null;
    }
}

File: javascript/src/task-service.js
var Task = require('./task');
var DeadlineToday = require('./deadline').DeadlineToday;

function TaskService() {
  this._projects = {};
  this._tasksDueToday = [];
}
TaskService.prototype = {

  _projects: null,
  _taskIdCounter: 0,
  _tasksDueToday: null,

  addProject: function(projectName) {
    return this._projects[projectName] = [];
  },

  addTaskToProject: function(projectName, taskName) {
    if (!this._projects[projectName]) {
      throw Error;
    }
    var task = new Task(taskName);
    this._projects[projectName].push(task);
    return task.id;
  },

  addDeadlineToTask: function(taskId, deadline) {
    if (deadline instanceof DeadlineToday) {
      this._tasksDueToday.push(taskId);
    }
  },

  showTasksDueToday: function() {
    return this._tasksDueToday.map(this.findTaskById, this);
  },

  findTaskById: function(taskId) {
    var tasks, projectName, task;
    tasks = [];
    for (projectName in this._projects) {
      tasks = tasks.concat((fn$.call(this)));
    }
    return tasks[0] || null;
    function fn$() {
      var i$, ref$, len$, results$ = [];
      for (i$ = 0, len$ = (ref$ = this._projects[projectName]).length; i$ < len$; ++i$) {
        task = ref$[i$];
        if (task.id === taskId) {
          results$.push(task);
        }
      }
      return results$;
    }
  },

  findAllProjects: function() {
    return Object.keys(this._projects);
  },

  findAllTasksForProject: function(projectName) {
    return this._projects[projectName] || null;
  }
};

module.exports = TaskService;
